### DISCUSSION TITLE
SelfHosted Twinpack Server

### DISCUSSION THREAD

**Discussion by ThatsFokus** [Zeugwerk/Twinpack#134] [Q&A]
If I wanted to host a private server , would I need a specific application or would a normal nuget server already work out of the box?

Should I need a specific apllication, is there a repository where I could find the needed files?

**Comment by beslst** [comment]
We have a commercial product called 'Twinpack Server', which is optimized towards a http://releaseflow.org/ if this is not needed, any artifact server supporting the nuget protocol works, i.e. Nexus, Artifactory...

We didn’t update the documentation about this topic yet (still a todo), but the basic steps are to

- setup an artifact server with a nuget feed
- upload packages to it “normally”, i.e. with nuget deploy
- nuget packages should have the tag “plc-library” on it (atm this is not needed, but we will make it mandatory in a future update) and either contain a .library file or an msi installer, which in turn contains a .library file
- In Twinpack you can use the configuration dialog to add your private feed 

if your company would like some general consulting on CI/CD, feel free to contact us on info@zeugwerk.at

**Comment by llagit** [comment]
Hi, is it possible to connect from XAE / Visula studio to a nuget feed of Azure DevOps? I keep getting a login error even if I am using correctly the PAT without nay username.

**Reply by beslst** [reply]
this should work. If you don't have to authenticate at this server, it is enough to put in the url of the feed you don't have to login. Type in the feed url and set the repository type to "Nuget Repository", close the configuration and the packages should appear. At least one package has to be available in the feed, otherwise there will be a connection error.
